test/DGmethods_old/Euler/RTB_1DIMEX.jl

```diff
@@ -0,0 +1,427 @@
+
+  θ0::DFloat = 303
+  π_exner    = 1 - gravity / (c_p * θ0) * y # exner pressure
+  ρ          = p0 / (R_gas * θ0) * (π_exner)^ (c_v / R_gas) # density
+
+  P          = p0 * (R_gas * (ρ * θ0) / p0) ^(c_p/c_v) # pressure (absolute)
+  T          = P / (ρ * R_gas) # temperature
+  u⃗          = SVector(-zero(DFloat), -zero(DFloat), -zero(DFloat))
+  # energy definitions
+  e_kin = u⃗' * u⃗ / 2
+  e_pot = gravity * y
+  e_int = cv_d *  T
+  ρe    = ρ * (e_int + e_kin + e_pot)
+  ρ, ρe
+end
+
+# Initial Condition
+function rising_bubble!(dim, Q, t, x, y, z, aux)
+  DFloat          = eltype(Q)
+  R_gas::DFloat   = R_d
+  c_p::DFloat     = cp_d
+  c_v::DFloat     = cv_d
+  p0::DFloat      = MSLP
+  gravity::DFloat = grav
+  # perturbation parameters for rising bubble
+
+  r⃗ = SVector(x, y, z)
+  r⃗_center = SVector(500, 260, 500)
+  distance = norm(@view (r⃗ - r⃗_center)[1:dim])
+
+  θ0::DFloat  = 303
+  θ_c::DFloat = 1 // 2
+  Δθ::DFloat  = -zero(DFloat)
+  a::DFloat   =  50
+  s::DFloat   = 100
+  if distance <= a
+    Δθ = θ_c
+  elseif distance > a
+    Δθ = θ_c * exp(-(distance - a)^2 / s^2)
+  end
+  θ       = θ0 + Δθ # potential temperature
+  π_exner = 1 - gravity / (c_p * θ) * y # exner pressure
+  ρ       = p0 / (R_gas * θ) * (π_exner)^ (c_v / R_gas) # density
+
+  P       = p0 * (R_gas * (ρ * θ) / p0) ^(c_p/c_v) # pressure (absolute)
+  T       = P / (ρ * R_gas) # temperature
+  u⃗       = SVector(-zero(DFloat), -zero(DFloat), -zero(DFloat))
+  # energy definitions
+  e_kin = u⃗' * u⃗ / 2
+  e_pot = gravity * y
+  e_int = cv_d *  T
+  ρe    = ρ * (e_int + e_kin + e_pot)
+  ρu⃗    = ρ * u⃗
+
+  @inbounds ρ0, ρe0 = aux[_a_ρ0], aux[_a_ρe0]
+  @inbounds Q[_δρ], Q[_δρe] = ρ-ρ0, ρe-ρe0
+  @inbounds Q[_ρu⃗] = ρu⃗
+end
+
+# {{{ Linearization
+@inline function lin_eulerflux!(F, Q, aux, t)
+  F .= -zero(eltype(Q))
+  @inbounds begin
+    DFloat = eltype(Q)
+    γ::DFloat = γ_exact # FIXME: Remove this for some moist thermo approach
+
+    δρ, δρe = Q[_δρ], Q[_δρe]
+    ρu⃗ = SVector(Q[_ρu], Q[_ρv], Q[_ρw])
+
+    ρ0, ρe0 = aux[_a_ρ0], aux[_a_ρe0]
+    ϕ = aux[_a_ϕ]
+
+    ρinv0 = 1 / ρ0
+    e0 = ρinv0 * ρe0
+
+    P0 = (γ-1)*(ρe0 - ρ0 * ϕ)
+    δP = (γ-1)*(δρe - δρ * ϕ)
+
+    p0 = ρinv0 * P0
+
+    F[:, _δρ ] = ρu⃗
+    F[:, _ρu⃗ ] = δP * I + @SMatrix zeros(3,3)
+    F[:, _δρe] = (e0 + p0) * ρu⃗
+  end
+end
+
+@inline function wavespeed_linear(n, Q, aux, t)
+  @inbounds begin
+    DFloat = eltype(Q)
+    γ::DFloat = γ_exact # FIXME: Remove this for some moist thermo approach
+
+    ρ0, ρe0, ϕ = aux[_a_ρ0], aux[_a_ρe0], aux[_a_ϕ]
+    ρinv0 = 1 / ρ0
+    P0 = (γ-1)*(ρe0 - ρ0 * ϕ)
+
+    sqrt(ρinv0 * γ * P0)
+  end
+end
+
+@inline function lin_source!(S, Q, aux, t)
+  S .= 0
+  lin_source_geopotential!(S, Q, aux, t)
+end
+@inline function lin_source_geopotential!(S, Q, aux, t)
+  @inbounds begin
+    δρ = Q[_δρ]
+    ∇ϕ = SVector(aux[_a_ϕ_x], aux[_a_ϕ_y], aux[_a_ϕ_z])
+    S[_ρu] -= δρ * ∇ϕ[1]
+    S[_ρv] -= δρ * ∇ϕ[2]
+    S[_ρw] -= δρ * ∇ϕ[3]
```

Comment:
Does it make a substantial difference performance-wise whether this buoyancy term is included in the linear solve or not? (I think it makes sense to include it here for a GCM configuration, I’m just wondering how much, if any, overhead it creates.)

Reply:
Good question.

If we are using a direct solver then I think the performance penalty is only in the initialization.

If we are using an iterative method to solve the system, there has to be some performance penalty because we are loading more data but we would have to measure to find out how much (I am guessing not much).  Also I don't know the impact it has on the number of iterations of the iterative solver, maybe @fxgiraldo has some experience.